java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,261 @@
+  }
+
+  public boolean isConstrained(ProgramState ps, SymbolicValue val) {
+    return ps.constraints.containsKey(val);
+  }
+
+  public Pair<ProgramState, ProgramState> assumeDual(ProgramState programState, Tree condition) {
+    // FIXME condition value should be evaluated to determine if it is worth exploring this branch. This should probably be done in a
+    // dedicated checker.
+    condition = skipTrivial(condition);
+    switch (condition.kind()) {
+      case INSTANCE_OF: {
+        InstanceOfTree instanceOfTree = (InstanceOfTree) condition;
+        SymbolicValue exprValue = eval(programState, instanceOfTree.expression());
+        if (isNull(programState, exprValue)) {
+          return new Pair<>(programState, null);
+        }
+        // if instanceof is true then we know for sure that expression is not null.
+        return new Pair<>(programState, setConstraint(programState, exprValue, NullConstraint.NOT_NULL));
+      }
+      case EQUAL_TO: {
+        BinaryExpressionTree equalTo = (BinaryExpressionTree) condition;
+        SymbolicValue lhs = eval(programState, equalTo.leftOperand());
+        SymbolicValue rhs = eval(programState, equalTo.rightOperand());
+        if (isNull(programState, lhs)) {
+          ProgramState stateNull = setConstraint(programState, rhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, rhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        } else if (isNull(programState, rhs)) {
+          ProgramState stateNull = setConstraint(programState, lhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, lhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        }
+        break;
+      }
+      case NOT_EQUAL_TO: {
+        BinaryExpressionTree notEqualTo = (BinaryExpressionTree) condition;
+        SymbolicValue lhs = eval(programState, notEqualTo.leftOperand());
+        SymbolicValue rhs = eval(programState, notEqualTo.rightOperand());
+        if (isNull(programState, lhs)) {
+          ProgramState stateNull = setConstraint(programState, rhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, rhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNull, stateNotNull);
+        } else if (isNull(programState, rhs)) {
+          ProgramState stateNull = setConstraint(programState, lhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, lhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNull, stateNotNull);
+        }
+        break;
+      }
+      case LOGICAL_COMPLEMENT:
+        return assumeDual(programState, ((UnaryExpressionTree) condition).expression()).invert();
+      case CONDITIONAL_OR:
+      case CONDITIONAL_AND:
+        // this is the case for branches such as "if (lhs && rhs)" and "if (lhs || rhs)"
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) This block of commented-out lines of code should be removed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)